File: app/schemas/draft.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class CreateDraftReq(BaseModel):
    title: str = Field(default="未命名", max_length=255)
    workflow_type: str = Field(default="novel", max_length=50)
    draft_group_id: str | None = Field(default=None, description="UUID string for grouping related drafts")


class UpsertDraftReq(BaseModel):
    title: str = Field(default="未命名", max_length=255)
    workflow_type: str = Field(default="novel", max_length=50)
    current_step: str = Field(default="prompt", max_length=50)
    step_data: dict = Field(default_factory=lambda: {"schema_version": 1})


class UpdateDraftReq(BaseModel):
    title: str | None = Field(None, max_length=255)
    status: str | None = Field(None, max_length=20)
    current_step: str | None = Field(None, max_length=50)
    step_data: dict | None = None


class DraftResp(BaseModel):
    id: UUID
    title: str
    workflow_type: str
    draft_group_id: UUID | None = None
    status: str
    current_step: str
    step_data: dict
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DraftListItem(BaseModel):
    id: UUID
    title: str
    workflow_type: str
    draft_group_id: UUID | None = None
    status: str
    current_step: str
    updated_at: datetime

    model_config = {"from_attributes": True}


class GenerateChapterReq(BaseModel):
    gen_model: str | None = Field(
        None, description="Model override, e.g. 'claude::claude-sonnet-4-20250514'"
    )
    chapter_num: int | None = Field(
        None, description="Specific chapter number to regenerate (1-based). If omitted, generates the next chapter."
    )


class VolumeReviewReport(BaseModel):
    review_text: str
    decision: str
    parsed_decision: str
    analysis_summary: str = ""


class VolumeReviewResp(BaseModel):
    volume_review_report: VolumeReviewReport
    chapter_count: int
    total_chapters: int
    volume_2_outline: str | None = None
    revised_chapters: list[dict] | None = None


class FinalReviewResp(BaseModel):
    report: str
    chapter_count: int
    total_chapters: int
    revised_chapters: list[dict] | None = None


class GenerateChapterResp(BaseModel):
    chapter_num: int
    chapter_title: str
    chapter_content: str
    total_chapters: int
    draft: DraftResp
    quality_check_needed: bool = False
    volume_review: VolumeReviewResp | None = None
    final_review: FinalReviewResp | None = None


class VolumeReviewDecisionReq(BaseModel):
    decision: str = Field(
        ..., description="One of: 续写第二卷, 修改后继续, 收束结局"
    )
    apply_revisions: bool = Field(
        default=False, description="Whether to apply chapter revisions from the review"
    )


class SubmitVolumeDecisionResp(BaseModel):
    message: str
    volume_2_outline: str | None = None
    new_total_chapters: int
    draft: DraftResp


class FinalNovelDecisionReq(BaseModel):
    apply_revisions: bool = Field(
        default=False, description="Whether to apply revisions from the final review"
    )
    mark_complete: bool = Field(
        default=True, description="Whether to mark the novel as complete"
    )


class FinalNovelDecisionResp(BaseModel):
    message: str
    draft: DraftResp

File: client/src/types/draft.ts
import type { SearchResultItem } from './novel';

// ── Novel-specific step data ──
export interface NovelDraftStepData {
  schema_version?: number;
  keywords?: string;
  customPrompt?: string;
  genModel?: string;
  references?: SearchResultItem[];
  analysis?: string;
  outlineText?: string;
  volumeOutlineText?: string;
  characterRulesText?: string;
  novelContent?: string;
  chapters?: { title: string; content: string }[];
  totalChapters?: number;
  generateMode?: 'batch' | 'interactive';
  qualityReport?: string | null;
  qualityRevisions?: ChapterRevision[];
  qualityRevisionsSummary?: string;
  // Volume review state (interactive mode)
  volumeReviewState?: 'pending_review' | 'executing_v2' | 'executing_closing' | 'pending_final_review' | 'completed' | null;
  volumeReviewReport?: VolumeReviewReport | null;
  volumeReviewDecision?: string | null;
  volume2Outline?: string | null;
  volume2ChaptersWritten?: number;
  closingChaptersWritten?: number;
  // Final review (when v2/closing arc completes)
  finalReviewReport?: string | null;
  finalReviewRevisions?: ChapterRevision[];
  finalReviewApplied?: boolean;
}

export type DraftStepData = Record<string, unknown>;

export interface Draft {
  id: string;
  title: string;
  workflow_type: string;
  draft_group_id?: string;
  status: 'in_progress' | 'completed';
  current_step: string;
  step_data: DraftStepData;
  created_at: string;
  updated_at: string;
}

export interface DraftListItem {
  id: string;
  title: string;
  workflow_type: string;
  draft_group_id?: string;
  status: string;
  current_step: string;
  updated_at: string;
}

export interface CreateDraftReq {
  title?: string;
  workflow_type: string;
  draft_group_id?: string;
}

export interface UpsertDraftReq {
  title?: string;
  workflow_type: string;
  current_step: string;
  step_data: DraftStepData;
}

export interface UpdateDraftReq {
  title?: string;
  status?: string;
  current_step?: string;
  step_data?: DraftStepData;
}

// Novel-specific generate-chapter types
export interface GenerateChapterReq {
  gen_model?: string;
  chapter_num?: number;
}

export interface VolumeReviewReport {
  review_text: string;
  decision: string;
  parsed_decision: string;
  analysis_summary: string;
}

export interface VolumeReviewResp {
  volume_review_report: VolumeReviewReport;
  chapter_count: number;
  total_chapters: number;
  volume_2_outline?: string | null;
  revised_chapters?: { title: string; content: string }[] | null;
}

export interface GenerateChapterResp {
  chapter_num: number;
  chapter_title: string;
  chapter_content: string;
  total_chapters: number;
  draft: Draft;
  quality_check_needed?: boolean;
  volume_review?: VolumeReviewResp | null;
  final_review?: FinalReviewResp | null;
}

export interface FinalReviewResp {
  report: string;
  chapter_count: number;
  total_chapters: number;
  revised_chapters?: { chapter_index: number; title: string; content: string }[] | null;
}

export interface VolumeReviewDecisionReq {
  decision: '续写第二卷' | '修改后继续' | '收束结局';
  apply_revisions?: boolean;
}

export interface SubmitVolumeDecisionResp {
  message: string;
  volume_2_outline?: string | null;
  new_total_chapters: number;
  draft: Draft;
}

export interface ChapterRevision {
  chapter_index: number;
  title: string;
  content: string;
}

export interface AnalyzeChaptersResp {
  report: string;
  revisions: ChapterRevision[];
  revisions_summary: string;
}

export interface FinalNovelDecisionReq {
  apply_revisions?: boolean;
  mark_complete?: boolean;
}

export interface FinalNovelDecisionResp {
  message: string;
  draft: Draft;
}
